# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: layout/svg/SVGIntegrationUtils.h
# Commit: bdb29ecb2b16
# Full Hash: bdb29ecb2b16458ca5c84594a5c9a0a6febc6249
# Author: Matt Woodrow <mwoodrow@mozilla.com>
# Date: 2021-05-16 09:17:48
# Regressor Bug: 1540737
# File Overlap Count: 1
# Description:
#   Bug 1540737 - Allow filter painting to take a callback for painting children. r=miko
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D114670
# ==============================================================================

diff -r f54a65e66247 -r bdb29ecb2b16 layout/svg/SVGIntegrationUtils.h
--- a/layout/svg/SVGIntegrationUtils.h	Sat May 15 22:16:48 2021 +0000
+++ b/layout/svg/SVGIntegrationUtils.h	Sat May 15 22:38:43 2021 +0000
@@ -204,9 +204,26 @@
   static bool IsMaskResourceReady(nsIFrame* aFrame);
 
   /**
+   * Paint the frame contents.
+   * SVG frames will have had matrix propagation set to false already.
+   * Non-SVG frames have to do their own thing.
+   * The caller will do a Save()/Restore() as necessary so feel free
+   * to mess with context state.
+   * The context will be configured to use the "user space" coordinate
+   * system.
+   * @param aDirtyRect the dirty rect *in user space pixels*
+   * @param aTransformRoot the outermost frame whose transform should be taken
+   *                       into account when painting an SVG glyph
+   */
+  using SVGFilterPaintCallback = std::function<void(
+      gfxContext& aContext, nsIFrame* aTarget, const gfxMatrix& aTransform,
+      const nsIntRect* aDirtyRect, image::imgDrawingParams& aImgParams)>;
+
+  /**
    * Paint non-SVG frame with filter and opacity effect.
    */
-  static void PaintFilter(const PaintFramesParams& aParams);
+  static void PaintFilter(const PaintFramesParams& aParams,
+                          const SVGFilterPaintCallback& aCallback);
 
   /**
    * Build WebRender filters for a frame with CSS filters applied to it.
@@ -268,6 +285,13 @@
    * For SVG frames, this returns a zero offset.
    */
   static nsPoint GetOffsetToBoundingBox(nsIFrame* aFrame);
+
+  /**
+   * The offset between the reference frame and the bounding box of the
+   * target frame in device units.
+   */
+  static gfxPoint GetOffsetToUserSpaceInDevPx(nsIFrame* aFrame,
+                                              const PaintFramesParams& aParams);
 };
 
 }  // namespace mozilla